Parser/pegen/pegen.c

```diff
@@ -383,48 +383,61 @@ tokenizer_error(Parser *p)
 }
 
 void *
-_PyPegen_raise_error(Parser *p, PyObject *errtype, int with_col_number, const char *errmsg, ...)
+_PyPegen_raise_error(Parser *p, PyObject *errtype, int with_col_number,
+                     const char *errmsg, ...)
+{
+    Token *t = p->known_err_token != NULL ? p->known_err_token : p->tokens[p->fill - 1];
+    int col_offset;
+    if (t->col_offset == -1) {
+        col_offset = Py_SAFE_DOWNCAST(p->tok->cur - p->tok->buf,
+                                      intptr_t, int);
+    } else {
+        col_offset = t->col_offset + 1;
+    }
+
+    va_list va;
+    va_start(va, errmsg);
+    _PyPegen_raise_error_known_location(p, errtype, with_col_number, t->lineno,
+                                        col_offset, errmsg, va);
+    va_end(va);
+
+    return NULL;
+}
+
+
+void *
+_PyPegen_raise_error_known_location(Parser *p, PyObject *errtype,
+                                   int with_col_number,
+                                   int lineno, int col_offset,
+                                   const char *errmsg, va_list va)
 {
     PyObject *value = NULL;
     PyObject *errstr = NULL;
-    PyObject *loc = NULL;
+    PyObject *error_line = NULL;
     PyObject *tmp = NULL;
-    Token *t = p->known_err_token != NULL ? p->known_err_token : p->tokens[p->fill - 1];
-    Py_ssize_t col_number = !with_col_number;
-    va_list va;
     p->error_indicator = 1;
 
-    va_start(va, errmsg);
     errstr = PyUnicode_FromFormatV(errmsg, va);
-    va_end(va);
     if (!errstr) {
         goto error;
     }
 
     if (p->start_rule == Py_file_input) {
-        loc = PyErr_ProgramTextObject(p->tok->filename, t->lineno);
+        error_line = PyErr_ProgramTextObject(p->tok->filename, lineno);
     }
 
-    if (!loc) {
-        loc = get_error_line(p->tok->buf, p->start_rule == Py_file_input);
+    if (!error_line) {
+        error_line = get_error_line(p->tok->buf, p->start_rule == Py_file_input);
     }
 
-    if (loc && with_col_number) {
-        int col_offset;
-        if (t->col_offset == -1) {
-            col_offset = Py_SAFE_DOWNCAST(p->tok->cur - p->tok->buf,
-                                          intptr_t, int);
-        } else {
-            col_offset = t->col_offset + 1;
-        }
-        col_number = byte_offset_to_character_offset(loc, col_offset);
-    }
-    else if (!loc) {
+    int col_number = byte_offset_to_character_offset(error_line, col_offset);
+
+    if (!error_line || (!with_col_number && p->start_rule != Py_file_input)) {
         Py_INCREF(Py_None);
-        loc = Py_None;
+        error_line = Py_None;
     }
 
-    tmp = Py_BuildValue("(OiiN)", p->tok->filename, t->lineno, col_number, loc);
+    tmp = Py_BuildValue("(OiiN)", p->tok->filename, lineno, col_number, error_line);
```

Comment:
Honestly I think it's perverse to set `error_line` to None in order to prevent the caret, since this not only prevents showing the caret but also prevents showing the text of the line in error at all. But you have the line number, so you should show the text of the line.

What I would expect is creating a tuple `(filename, lineno, None, error_line)` where `error_line` is the text of the line. (This is what I have been hinting at previously.)

Reply:
> since this not only prevents showing the caret but also prevents showing the text of the line in error at all. But you have the line number, so you should show the text of the line.

The old parser behaves that way. For example:
```
╰─ python3.8                                                     
Python 3.8.1+ (heads/3.8:33e033da3c, Jan 10 2020, 01:13:05) 
[GCC 9.2.1 20191008] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 1 + 1 = 2
  File "<stdin>", line 1
SyntaxError: cannot assign to operator
```
It show no carret and no text.

By looking at the code, this indeed seems to be the intention of `Python/ast.c`.
https://github.com/python/cpython/blob/cd8295ff758891f21084a6a5ad3403d35dda38f7/Python/ast.c#L643
If it were up to me, I'd completely change that behavior in the new parser and _always_  show both the caret and the error text. Would that be okay, you think?

Reply:
That's different. (And the variable name `loc` doesn't help.) It replaces NULL with None if `PyErr_ProgramTextObject()` couldn't find the line, and with the old parser that is always the case with REPL input. (It has to do with reading the source file a second time to get the text.)

If you put this in a file the caret points to the start of the bad statement:
```
if 1: 1+1 = 1
```
Now the old parser gives:
```
~/cpython$ ./python.exe -X oldparser t.py
  File "/Users/guido/cpython/t.py", line 1
    if 1: 1+1 = 1
          ^
SyntaxError: cannot assign to operator
~/cpython$ 
```
Please read the rest of my comments carefully.